zebra-state/src/utxo.rs

```diff
@@ -18,15 +22,19 @@ pub struct Utxo {
     pub from_coinbase: bool,
 }
 
-#[cfg(test)]
-pub fn new_outputs(block: &block::Block) -> std::collections::HashMap<transparent::OutPoint, Utxo> {
-    use std::collections::HashMap;
-
-    let height = block.coinbase_height().expect("block has coinbase height");
-
+/// Compute an index of newly created transparent outputs, given a block and a
+/// list of precomputed transaction hashes.
+pub fn new_outputs(
+    block: &Block,
+    transaction_hashes: &[transaction::Hash],
```

Comment:
Minor: Shouldn't this be calculated inside the function? There is a (small) risk of calling this function with the incorrect `transaction_hashes` (or an empty slice).

Maybe this should be done later, but one idea might be to have a `Block::transaction_hashes(&self) -> impl Iterator<&Hash>` (or `transaction_ids`) method and a `Block::new_outputs` method. The latter could use the former method.

Reply:
My goal in this PR was to remove code duplication, while keeping the same interface. I didn't want to make major changes, but I do need to modify this code to get transparent validation working. So I didn't want any duplicate functions.

Here's some more context:

`new_outputs` is part of a design with tradeoffs between CPU usage, security, and a small risk of incorrect code.

Zebra calculates the transaction hashes exactly once for each block, either in the block verifier, or the checkpoint verifier. Then we re-use the cached hashes from the resulting `PreparedBlock` or `FinalizedBlock`, to save CPU usage.

For security reasons, we want to perform the block verifier checks in the following order:
1. Calculate transaction hashes
2. Check the merkle root (to bind transaction data to the block header, making sure they haven't been modified)
3. Do the other checks that use the transaction data

I've opened #2475 for further improvements.